pkgs/top-level/python-packages.nix

```diff
@@ -8469,7 +8469,17 @@ in {
     };
   };
 
-  pytorch = callPackage ../development/python-modules/pytorch { };
+  pytorch = callPackage ../development/python-modules/pytorch {
+    cudaSupport = pkgs.config.cudaSupport or false;
+  };
+
+  pytorchWithCuda = self.pytorch.override {
+    cudaSupport = true;
+  };
+
+  pytorchWithoutCuda = self.pytorch.override {
+    cudaSupport = false;
```

Comment:
I tried to build the package on a system without cuda and found that setting cudaSupport = false is actualy not enough. This way cuda library dependency still is being built. It disappeared then I set cudatoolkit and cudnn to null explicitly.

Reply:
I’d forgotten to `let`-bind `cudatoolkit_joined`, `cudaStub`, and `cudaStubEnv`. Does this work better now?

Reply:
@andersk looks like I messed with history. I actually tried my old version and stumbled upon issue mentioned here. https://github.com/NixOS/nixpkgs/pull/32438#issuecomment-378405683

Still, yours commit 7a851ad4c fails for me with the message:
```
nix-shell -p python36Packages.pytorch --show-trace                                                                                  
error: while evaluating the attribute ‘buildInputs’ of the derivation ‘shell’ at /opt/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:98:11:
while evaluating the attribute ‘patches’ of the derivation ‘python3.6-pytorch-0.3.1’ at /opt/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:98:11:
while evaluating anonymous function at /opt/nixpkgs/pkgs/build-support/fetchpatch/default.nix:8:1, called from /opt/nixpkgs/pkgs/development/python-modules/pytorch/default.nix:41:6:
anonymous function at /opt/nixpkgs/pkgs/build-support/fetchurl/default.nix:38:1 called with unexpected argument ‘extraPrefix’, at /opt/nixpkgs/pkgs/build-support/fetchpatch/default.nix:10:1
```

Reply:
You need to merge with master or release-18.03 or something.

Reply:
I merged your branch to fresh master at e1dee4ef as you suggested. Builds for me now, but fails on testing. I run this on ancient P6200 processor, probably that's the reason.
```
$ nix-shell -p python36Packages.pytorchWithoutCuda
<...>
installing
/tmp/nix-build-python3.6-pytorch-0.3.1.drv-0/source/dist /tmp/nix-build-python3.6-pytorch-0.3.1.drv-0/source
Processing ./torch-0.3.1b0-cp36-cp36m-linux_x86_64.whl
Requirement already satisfied: numpy in /nix/store/q61pr9l2l31hqrkan0m4zd0wqv85cmnp-python3.6-numpy-1.14.1/lib/python3.6/site-packages (from torch==0.3.1b0)
Requirement already satisfied: pyyaml in /nix/store/rp8vvvf15fni2dxn1p46hv6ly3k0xca3-python3.6-PyYAML-3.12/lib/python3.6/site-packages (from torch==0.3.1b0)
Installing collected packages: torch
Successfully installed torch-0.3.1b0
/tmp/nix-build-python3.6-pytorch-0.3.1.drv-0/source
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/_thnn/_THNN.cpython-36m-x86_64-linux-gnu.so
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/lib/libTHNN.so.1
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/lib/libshm.so
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/lib/libTHS.so.1
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/lib/torch_shm_manager
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/lib/libATen.so.1
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/lib/libTH.so.1
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/_dl.cpython-36m-x86_64-linux-gnu.so
shrinking /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib/python3.6/site-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
strip is /nix/store/fzcs0fn6bb04m82frhlb78nc03ny3w55-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1/lib 
patching script interpreter paths in /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1
checking for references to /tmp/nix-build-python3.6-pytorch-0.3.1.drv-0 in /nix/store/5v8kzkyk9k11icfkzf0vbf7vr3rpqgdw-python3.6-pytorch-0.3.1...
running install tests
/tmp/nix-build-python3.6-pytorch-0.3.1.drv-0/source/test /tmp/nix-build-python3.6-pytorch-0.3.1.drv-0/source
Running JIT tests
s.......s.s.s..s....s.....s......ss.........
----------------------------------------------------------------------
Ran 44 tests in 0.285s

OK (skipped=9)
Running torch tests
...test/run_test.sh: line 27: 29365 Illegal instruction     (core dumped) $PYCMD test_torch.py $@
builder for ‘/nix/store/4bs5qi4jpnslv3aqr8m8ffp19xy1bi9d-python3.6-pytorch-0.3.1.drv’ failed with exit code 132hell -p python36Packages.pytorchWithoutCuda
```

Reply:
ok. Builts with and without cuda on a more recent hardware.

Reply:
What does this section of your build log look like on your P6200?

```
-- Performing Test C_HAS_SSE1_1
-- Performing Test C_HAS_SSE1_1 - Success
-- Performing Test C_HAS_SSE2_1
-- Performing Test C_HAS_SSE2_1 - Success
-- Performing Test C_HAS_SSE3_1
-- Performing Test C_HAS_SSE3_1 - Failed
-- Performing Test C_HAS_SSE3_2
-- Performing Test C_HAS_SSE3_2 - Success
-- Performing Test C_HAS_SSE4_1_1
-- Performing Test C_HAS_SSE4_1_1 - Failed
-- Performing Test C_HAS_SSE4_1_2
-- Performing Test C_HAS_SSE4_1_2 - Success
-- Performing Test C_HAS_SSE4_2_1
-- Performing Test C_HAS_SSE4_2_1 - Failed
-- Performing Test C_HAS_SSE4_2_2
-- Performing Test C_HAS_SSE4_2_2 - Success
-- Performing Test C_HAS_AVX_1
-- Performing Test C_HAS_AVX_1 - Failed
-- Performing Test C_HAS_AVX_2
-- Performing Test C_HAS_AVX_2 - Success
-- Performing Test C_HAS_AVX2_1
-- Performing Test C_HAS_AVX2_1 - Failed
-- Performing Test C_HAS_AVX2_2
-- Performing Test C_HAS_AVX2_2 - Success
-- Performing Test CXX_HAS_SSE1_1
-- Performing Test CXX_HAS_SSE1_1 - Success
-- Performing Test CXX_HAS_SSE2_1
-- Performing Test CXX_HAS_SSE2_1 - Success
-- Performing Test CXX_HAS_SSE3_1
-- Performing Test CXX_HAS_SSE3_1 - Failed
-- Performing Test CXX_HAS_SSE3_2
-- Performing Test CXX_HAS_SSE3_2 - Success
-- Performing Test CXX_HAS_SSE4_1_1
-- Performing Test CXX_HAS_SSE4_1_1 - Failed
-- Performing Test CXX_HAS_SSE4_1_2
-- Performing Test CXX_HAS_SSE4_1_2 - Success
-- Performing Test CXX_HAS_SSE4_2_1
-- Performing Test CXX_HAS_SSE4_2_1 - Failed
-- Performing Test CXX_HAS_SSE4_2_2
-- Performing Test CXX_HAS_SSE4_2_2 - Success
-- Performing Test CXX_HAS_AVX_1
-- Performing Test CXX_HAS_AVX_1 - Failed
-- Performing Test CXX_HAS_AVX_2
-- Performing Test CXX_HAS_AVX_2 - Success
-- Performing Test CXX_HAS_AVX2_1
-- Performing Test CXX_HAS_AVX2_1 - Failed
-- Performing Test CXX_HAS_AVX2_2
-- Performing Test CXX_HAS_AVX2_2 - Success
-- SSE2 Found
-- SSE3 Found
-- AVX Found
-- AVX2 Found
```

Should we add `sse2Support`, `sse3Support`, `avxSupport`, `avx2Support` flags for better determinism?